File: registroUsuarios/PluginBiometrico.NET/src/PluginBiometrico.App/Ayudantes/ProbadorConexionServidor.cs
using System.Net.Http;
using System.Text.Json;
using PluginBiometrico.Core.Modelos;

namespace PluginBiometrico.App.Ayudantes;

/// <summary>
/// Verifica que el servidor PHP responda antes de guardar la configuración.
/// </summary>
public static class ProbadorConexionServidor
{
    private static readonly JsonSerializerOptions OpcionesJson = new()
    {
        PropertyNameCaseInsensitive = true
    };

    public static async Task<(bool Exito, string Mensaje)> ProbarAsync(ConfiguracionLocal config)
    {
        if (!ConstructorUrlsServidor.EsUrlValida(config.UrlHabilitarSensor))
        {
            return (false, "La URL del sensor no es válida. Use http:// o https://");
        }

        if (!ConstructorUrlsServidor.EsUrlValida(config.UrlApiRest))
        {
            return (false, "La URL de la API REST no es válida. Use http:// o https://");
        }

        if (string.IsNullOrWhiteSpace(config.IdUnicoPc))
        {
            return (false, "Indique el ID único de esta PC (token).");
        }

        using var http = new HttpClient { Timeout = TimeSpan.FromSeconds(12) };
        var token = Uri.EscapeDataString(config.IdUnicoPc.Trim());

        var urlSensor = $"{config.UrlHabilitarSensor.Trim()}" +
                        $"?timestamp=0&token={token}&ping=1" +
                        $"&_={DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}";

        var (exitoSensor, mensajeSensor) = await ProbarSensorAsync(http, urlSensor);
        if (exitoSensor)
        {
            return (true, mensajeSensor ?? "Conexión correcta con el sensor.");
        }

        var urlApi = $"{config.UrlApiRest.Trim()}?desde=0&hasta=0&token={token}";
        var (exitoApi, mensajeApi) = await ProbarApiRestAsync(http, urlApi);
        if (exitoApi)
        {
            return (true,
                "Conexión con el servidor de producción correcta (API REST). " +
                "Actualice HabilitarSensor.php en el servidor para habilitar la prueba directa del sensor.");
        }

        return (false, mensajeSensor ?? mensajeApi ?? "No se pudo conectar al servidor.");
    }

    private static async Task<(bool Exito, string? Mensaje)> ProbarSensorAsync(HttpClient http, string url)
    {
        try
        {
            using var request = CrearPeticionGet(url);
            using var response = await http.SendAsync(request);
            var cuerpo = await response.Content.ReadAsStringAsync();

            if (!response.IsSuccessStatusCode)
            {
                return (false, $"HabilitarSensor.php respondió con error HTTP {(int)response.StatusCode}.");
            }

            var comando = JsonSerializer.Deserialize<ComandoSensor>(cuerpo, OpcionesJson);
            if (comando is null)
            {
                return (false, "HabilitarSensor.php respondió, pero el JSON no es válido.");
            }

            return (true, $"Conexión correcta. El sensor responde (operación: {comando.Operacion ?? "reintentar"}).");
        }
        catch (TaskCanceledException)
        {
            return (false, null);
        }
        catch (HttpRequestException ex)
        {
            return (false, $"No se pudo conectar a HabilitarSensor.php: {ex.Message}");
        }
        catch (JsonException)
        {
            return (false, "HabilitarSensor.php respondió, pero no devolvió JSON válido.");
        }
    }

    private static async Task<(bool Exito, string? Mensaje)> ProbarApiRestAsync(HttpClient http, string url)
    {
        try
        {
            using var request = CrearPeticionGet(url);
            using var response = await http.SendAsync(request);
            var cuerpo = await response.Content.ReadAsStringAsync();

            if (!response.IsSuccessStatusCode)
            {
                return (false, $"UsuarioRestApi.php respondió con error HTTP {(int)response.StatusCode}.");
            }

            using var doc = JsonDocument.Parse(cuerpo);
            if (doc.RootElement.ValueKind != JsonValueKind.Array)
            {
                return (false, "UsuarioRestApi.php respondió, pero el JSON no es un arreglo válido.");
            }

            return (true, "Conexión correcta vía API REST.");
        }
        catch (TaskCanceledException)
        {
            return (false, "Tiempo de espera agotado. Verifique la URL de producción y que el servidor responda.");
        }
        catch (HttpRequestException ex)
        {
            return (false, $"No se pudo conectar: {ex.Message}");
        }
        catch (JsonException)
        {
            return (false, "UsuarioRestApi.php respondió, pero no devolvió JSON válido.");
        }
    }

    private static HttpRequestMessage CrearPeticionGet(string url)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.UserAgent.ParseAdd("PluginBiometrico/1.0");
        request.Headers.AcceptCharset.ParseAdd("UTF-8");
        return request;
    }
}

File: registroUsuarios/PluginBiometricoV4/src/Ayudantes/ProbadorConexionServidor.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package Ayudantes;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.JsonSyntaxException;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;

/**
 * Verifica que el servidor PHP responda antes de guardar la configuración,
 * replicando PluginBiometrico.App.Ayudantes.ProbadorConexionServidor del
 * plugin .NET (primero prueba HabilitarSensor.php con ping=1; si falla,
 * intenta UsuarioRestApi.php como respaldo).
 *
 * @author Mauricio Herrera
 */
public class ProbadorConexionServidor {

    public static final class Resultado {

        public final boolean exito;
        public final String mensaje;

        public Resultado(boolean exito, String mensaje) {
            this.exito = exito;
            this.mensaje = mensaje;
        }
    }

    private ProbadorConexionServidor() {
    }

    public static Resultado probar(String urlHabilitarSensor, String urlApiRest, String idUnicoPc) {
        if (!ConstructorUrlsServidor.esUrlValida(urlHabilitarSensor)) {
            return new Resultado(false, "La URL del sensor no es válida. Use http:// o https://");
        }
        if (!ConstructorUrlsServidor.esUrlValida(urlApiRest)) {
            return new Resultado(false, "La URL de la API REST no es válida. Use http:// o https://");
        }
        if (idUnicoPc == null || idUnicoPc.trim().isEmpty()) {
            return new Resultado(false, "Indique el ID único de esta PC (token).");
        }

        String token;
        try {
            token = URLEncoder.encode(idUnicoPc.trim(), "UTF-8");
        } catch (UnsupportedEncodingException ex) {
            token = idUnicoPc.trim();
        }

        String urlSensor = urlHabilitarSensor.trim() + "?timestamp=0&token=" + token + "&ping=1&_=" + System.currentTimeMillis();
        Resultado sensor = probarSensor(urlSensor);
        if (sensor.exito) {
            return sensor;
        }

        String urlApi = urlApiRest.trim() + "?desde=0&hasta=0&token=" + token;
        Resultado api = probarApiRest(urlApi);
        if (api.exito) {
            return new Resultado(true, "Conexión con el servidor correcta (API REST). "
                    + "Actualice HabilitarSensor.php en el servidor para habilitar la prueba directa del sensor.");
        }

        return new Resultado(false, sensor.mensaje != null ? sensor.mensaje : api.mensaje);
    }

    private static Resultado probarSensor(String url) {
        try {
            HttpURLConnection con = abrirConexion(url);
            int codigo = con.getResponseCode();
            String cuerpo = leerCuerpo(con, codigo);
            con.disconnect();
            if (codigo < 200 || codigo >= 300) {
                return new Resultado(false, "HabilitarSensor.php respondió con error HTTP " + codigo + ".");
            }
            JsonObject json = new JsonParser().parse(cuerpo).getAsJsonObject();
            String operacion = json.has("opc") && !json.get("opc").isJsonNull() ? json.get("opc").getAsString() : "reintentar";
            return new Resultado(true, "Conexión correcta. El sensor responde (operación: " + operacion + ").");
        } catch (JsonSyntaxException | IllegalStateException ex) {
            return new Resultado(false, "HabilitarSensor.php respondió, pero no devolvió JSON válido.");
        } catch (IOException ex) {
            return new Resultado(false, "No se pudo conectar a HabilitarSensor.php: " + ex.getMessage());
        }
    }

    private static Resultado probarApiRest(String url) {
        try {
            HttpURLConnection con = abrirConexion(url);
            int codigo = con.getResponseCode();
            String cuerpo = leerCuerpo(con, codigo);
            con.disconnect();
            if (codigo < 200 || codigo >= 300) {
                return new Resultado(false, "UsuarioRestApi.php respondió con error HTTP " + codigo + ".");
            }
            if (!cuerpo.trim().startsWith("[")) {
                return new Resultado(false, "UsuarioRestApi.php respondió, pero el JSON no es un arreglo válido.");
            }
            return new Resultado(true, "Conexión correcta vía API REST.");
        } catch (IOException ex) {
            return new Resultado(false, "No se pudo conectar: " + ex.getMessage());
        }
    }

    private static HttpURLConnection abrirConexion(String url) throws IOException {
        HttpURLConnection con = (HttpURLConnection) new URL(url).openConnection();
        con.setConnectTimeout(12000);
        con.setReadTimeout(12000);
        con.setRequestMethod("GET");
        con.setRequestProperty("User-Agent", "PluginBiometricoV4");
        con.setRequestProperty("Accept-Charset", "UTF-8");
        return con;
    }

    private static String leerCuerpo(HttpURLConnection con, int codigo) throws IOException {
        InputStream flujo = (codigo >= 200 && codigo < 300) ? con.getInputStream() : con.getErrorStream();
        if (flujo == null) {
            return "";
        }
        StringBuilder sb = new StringBuilder();
        try (BufferedReader in = new BufferedReader(new InputStreamReader(flujo))) {
            String linea;
            while ((linea = in.readLine()) != null) {
                sb.append(linea);
            }
        }
        return sb.toString();
    }
}
